Webhook server and framework for bot features with ACK tracker

aureleoules opened the pull request:
This PR adds a new summary comment that appears at the top of every (new) pull request.
It contains an ACK tracker that works as follows:

When a user comments or adds a review (edits and deletions are detected) or the PR author pushes to the PR, the bot will rescan every comment and review for ACKs.
An `ACK` can be a `(N)ACK`, `Concept (N)ACK`, `Approach (N)ACK`, `Code Review (N)ACK`, and `Stale ACK`.

For a raw `ACK` to be valid, a commit hash that matches the PR's head is expected (`ACK <commit sha>`).
If the commit hash no longer matches the PR's head, it goes stale. 
Concept ACKs, Code Review ACKs, Approach ACKs, and NACKs don't require a commit hash but one can be specified so that if the commit referenced no longer matches the head, the 'ACK' will be removed from the tracker automatically.
The case of the comment does not matter and takes into account ACK aliases such as re-ACK, reACK, utACK, crACK, etc.

Specifically, this PR:
- Adds a webhook server for receiving GitHub events of a repository
- Adds a framework for adding features to the Bot, with different event types
- Adds the summary comment feature

There is currently no secret-based webhook authentication but it can be done if this PR is Concept ACKed.

<details>
<summary>Summary comment</summary>

![image](https://user-images.githubusercontent.com/22493292/199021732-8b5c1c6e-6636-4d3a-9659-00b6ba3ea8cc.png)

</details>

<details>
<summary>ACK example</summary>

![image](https://user-images.githubusercontent.com/22493292/199023038-bc4c58ee-9f82-4e1c-b858-289f6def4791.png)

</details>

<details>
<summary>Stale ACK</summary>

![image](https://user-images.githubusercontent.com/22493292/199024155-a54de3ac-94df-4c64-a75f-96afd1364954.png)


</details>

Links in the `Reviewers` cell will automatically scroll down to the specific review.

kouloumos reviewed (commented):
Concept ACK
This gives another data point to someone glancing at the PRs page, giving more context than just solely the number of comments.
Although it goes without saying that not all ACKs are equal, I think this can give a quick overview on the sentiment of each PR without the need to get into each one. 

Furthermore, I think could be useful to also label for Stale ACKs, kind of how https://bitcoinacks.com/ does it. Cause reviewers don't come back right after a new commit is pushed to the PR, and with the current logic, you'll immediately lose the context that I praised above.

MarcoFalke commented on drahtbot/features/pr_label_acks.py:
```diff
@@ -0,0 +1,182 @@
+            break
+
+    if count > 0:
+        pr.add_to_labels(new_label)
```
This fails when the label doesn't exist. It might be best to ask first if people want the labels to exist. There was a previous discussion on IRC IIRC

aureleoules replied:
> This fails when the label doesn't exist

Interesting, the bot created the labels automatically for me. Maybe it's a permission issue?

> It might be best to ask first if people want the labels to exist. There was a previous discussion on IRC IIRC

Sure, should I create a GitHub issue on bitcoin/bitcoin to gain more visibility or just IRC? Also, would you have an archive link/timestamp for the previous discussion?

kouloumos replied:
> Also, would you have an archive link/timestamp for the previous discussion?

I found this discussion
https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358092;

aureleoules replied:
Thanks, interesting read. The way I see it, people didn't seem to be against it expect for @fanquake and @wumpus.

> [20:19](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358108) < wumpus> it's an interesting idea but i'm not really sure using labels for something like counting acks is a good idea, it really goes aside of the idea of labeling
> [20:20](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358109) < wumpus> there are already so many labels
> [20:21](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358110) < wumpus> ui-wise it's basically a more limited version of what bitcoinacks is doing but without the names
> [20:23](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358111) < wumpus> (and no- please don't create acked-by-wumpus acked-by-sipa etc labels xD)

> [22:04](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358117) < fanquake> I agree with wumpus, not sure about the value of “n ACKs” labels.
> [22:05](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358118) < fanquake> Also, please don’t have drahtbot start adding and removing ACK labels because some spam/unknown accounts drive by and drop ACKs on things. That’s just more noise.

I believe they were only talking about raw ACKs and not Concept/Approach/CR ACKs though. In case of raw ACKs I agree it doesn't say much about the conceptual state / review stage of a PR.

In terms of spam, I think spammers can and will find other ways to spam if they want to anyway.

Another issue I can see with labels is this: 
![image](https://user-images.githubusercontent.com/22493292/197180226-d7b156d6-3a08-457d-867c-a018b278b651.png)

There's a lot of them. So maybe instead of counting ACKs, there should only be one label per ACK type (except raw acks?).

MarcoFalke replied:
My idea was to have a label for an easy way to filter for pull request that have *no* current approvals, mostly as a way for reviewers to find stuff to conceptually/code review.

And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts). However, one downside would be that the merge script would then pick up that summary comment as a review comment itself.

aureleoules replied:
I see, in this case how about adding the label `Needs Concept ACK` if there isn't any `Concept ACK` comment? And removing it once there is. This means by default a PR will start with `Needs Concept ACK` label automatically. And once the PR is Concept ACKed add labels `Needs Approach ACK` and `Needs Code Review ACK`?

wumpus replied:
Wrong @wumpus

kouloumos replied:
> There's a lot of them. So maybe instead of counting ACKs, there should only be one label per ACK type (except raw acks?).

This is indeed noisy. I don't think that anyone will look for specific number of ACK. That said, following a "ACK(<number-of-acks>)" naming scheme seems better. If you could just sort them at the bottom of the list will probably solve this monstrosity — out of sight, out of mind. Also, "one label per ACK type" doesn't seem so useful. 

> And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts). However, one downside would be that the merge script would then pick up that summary comment as a review comment itself.

This was also part of the aforementioned IRC discussion (messages above the linked timestamp) that stood out to me. I have to test it first, but it seems that with a minor change [in the merge script](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/112deab9537e0bf41d74d6ac75b3bf693a26bb13/github-merge.py#L213), to check if the comment is by DrahtBot (`if (c['user']['login']] == "DrahtBot"): continue`), this could be solved. 

> I see, in this case how about adding the label `Needs Concept ACK` if there isn't any `Concept ACK` comment? And removing it once there is. This means by default a PR will start with `Needs Concept ACK` label automatically. And once the PR is Concept ACKed add labels `Needs Approach ACK` and `Needs Code Review ACK`?

This seems like a simpler, less controversial step to this labelling direction, which may help to boost interaction with PRs. Although the problem of "not all ACKs are equal" still exists, I think that this could be a nice experiment. I don't believe that there is need for multiple flags, `Needs Concept ACK` seems enough.

Note that right now there is a [Needs Conceptual Review](https://github.com/bitcoin/bitcoin/issues?q=is%3Aopen+is%3Apr+label%3A%22Needs+Conceptual+Review%22) label that is used in a different way — AFAIU is added when different opinions start to emerge regarding the change.

aureleoules replied:
Another issue I see with adding the `Needs Concept ACK` on every PR until someone comments `Concept ACK` is that it may be redundant on some PRs such as bug fixes that may have already been Concept ACKed in the corresponding issue.

There's a lot of PRs that have been merged without any Concept ACK comment: https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+NOT+%22Concept+ACK%22+in%3Acomment

MarcoFalke replied:
> This was also part of the aforementioned IRC discussion (messages above the linked timestamp) that stood out to me. I have to test it first, but it seems that with a minor change [in the merge script](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/112deab9537e0bf41d74d6ac75b3bf693a26bb13/github-merge.py#L213), to check if the comment is by DrahtBot (if (c['user']['login']] == "DrahtBot"): continue), this could be solved.

An easier way might be to just call them "reviews" instead of "ACKs", to avoid triggering the merge script.

aureleoules replied:
An idea I just had:
Drahtbot could create a summary comment when a PR is created, so it will appear at the very top and could be live-updated with the current ACKs/Approach ACKs/CR ACKs/etc. It would then be easy to glance at the state of the PR simply by reading this comment.

kouloumos replied:
>  it may be redundant on some PRs such as bug fixes that may have already been Concept ACKed in the corresponding issue.

Those PRs most of the times linked to the issue using [the supported keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). Maybe skipping the `Needs Concept ACK` stage if this is true. This does not apply to PRs that open as follow-up (maybe a heuristic could also apply here — 'followup' on the title?) and with PRs that open with a comment on another PR as reasoning (which I am not sure that is considered Concept ACK).

> There's a lot of PRs that have been merged without any Concept ACK comment: https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+NOT+%22Concept+ACK%22+in%3Acomment

I thought that ACK implies Concept ACK, no? That simplifies the logic. 

> An idea I just had: Drahtbot could create a summary comment when a PR is created, so it will appear at the very top and could be live-updated with the current ACKs/Approach ACKs/CR ACKs/etc. It would then be easy to glance at the state of the PR simply by reading this comment.

I don't think that this adds so much value regarding the "getting context on the state of a PR with a quick glance" use-case, because you are already in the PR page when watching this. 
That said, it's not a bad idea and I think that's what MarcoFalke said [above](https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1001687419): "And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts)". 

Adding to that, inspired by this part of the referenced IRC discussion
> [17:34](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358043) < jeremyrubin> when drahbot comments with the lists of PRs conflicting, a label that shows when the PR was last updated
...
[17:35](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358047) < jeremyrubin> it might be convenient to show if it's something that's stale or not

I think that displaying the current status of a conflicting PR (through that summary comment) might push more people reviewing that PR to go and have a look. The implementation of this might become easier if each PR's summary contained the live-updated ACKs status (calculate once, parsed multiple times?).

MarcoFalke replied:
Note: The labels of a PR (such as "Needs rebase", "Needs benchmark", ... ) are already shown when hovering over the link

aureleoules replied:
> Those PRs most of the times linked to the issue using [the supported keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). Maybe skipping the Needs Concept ACK stage if this is true. This does not apply to PRs that open as follow-up (maybe a heuristic could also apply here — 'followup' on the title?) and with PRs that open with a comment on another PR as reasoning (which I am not sure that is considered Concept ACK).

There are a lot of edge cases like this so maybe they should be addressed in a follow-up of this PR if this one ever gets merged. But I agree it might be possible to address these cases.

> I thought that ACK implies Concept ACK, no? That simplifies the logic.

Right, that could work then.

> I don't think that this adds so much value regarding the "getting context on the state of a PR with a quick glance" use-case, because you are already in the PR page when watching this.

Could be useful for large PRs with hundreds of comments. You wouldn't have to scroll all the way down, only read the first comment for a preview of the state of the PR I presume. Could also include in the summary comment whether there are unaddressed review comments for the latest commits or not.

I also agree with adding the last update, stale status, and conflicting PRs in the top summary comment.

aureleoules replied:
Also, this new top comment summary would only apply for new pulls because it needs to be at the top (for it to be useful, I think).

aureleoules replied:
Here's what the summary comment could look like. This was manually written, there's no code behind it yet.

![image](https://user-images.githubusercontent.com/22493292/198647218-dab2c166-411f-4fe2-8069-196ad80d454a.png)
I think this could be nice to have, and could also replace the current scripts so that conflicting PRs are computed the moment the PR is opened.

MarcoFalke replied:
Yeah, no objection. Just some nits:

* Not sure if it makes sense to differentiate reviews. Everyone uses their own description: review ACK, code ACK, ACK, utACK, tACK, tested ACK, cr ACK, ...
* Might also count NACKs?

> conflicting PRs are computed the moment the PR is opened.

It takes a few minutes to calculate them, so I think this is unrelated in any case.

aureleoules replied:
> Not sure if it makes sense to differentiate reviews. Everyone uses their own description: review ACK, code ACK, ACK, utACK, tACK, tested ACK, cr ACK, ...
    Might also count NACKs?

Yes, I plan to include all types of ACKs, this was just an example.
Also, I currently parse ACKs with aliases such as: `Code Review ACK`, `CR ACK`, `crACK`, `re-crACK` etc will all be included in the `Code Review ACK` cell, same for `tACK`, `utACK`, `ACK`, `reACK`, `re-ACK`.

MarcoFalke replied:
Ok, maybe ask other devs if anyone is objecting/liking this, since it requires a summary comment on *all* pulls.

kouloumos replied:
Looks good, I agree with both of MarcoFalke's points. I think to differentiate between Concept ACKs and the rest is enough. If you go this way, you could just have "Concept ACKs",  "Reviews", "Stale Reviews" ([as also mentioned above](https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1005508053)) and "NACKs".

I assume clicking on a reviewer's name takes you to their review?

aureleoules replied:
> Looks good, I agree with both of MarcoFalke's points. I think to differentiate between Concept ACKs and the rest is enough. If you go this way, you could just have "Concept ACKs", "Reviews", "Stale Reviews" (https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1005508053) and "NACKs".

Interesting, I would actually prefer to list different types of ACKs, which are: (N)ACKs, Concept (N)ACKs, Approach (N)ACKs, Code Review ACKs, NACKs, and Stale ACKs. I would omit them from the table if the count is 0 though.

It gives a bit more details on the review stage of the PR IMO, and whether it is ready for merge soon or not.

<details>
<summary>Examples</summary>


| ACKs | Count | Reviewers |
| --- | ----- | --------- |
| **NACKs**  | 1 | [aureleoules](https://github.com/aureleoules) |
| **ACKs** | 3 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [kouloumos](https://github.com/MarcoFalke) |
| Concept ACKs | 2 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) |
| Approach ACKs  | 1 | [aureleoules](https://github.com/aureleoules) |
| Code Review ACKs  | 1 | [Satoshi](https://github.com/satoshi) |
| _Stale_ ACKs | 1 | [Satoshi](https://github.com/satoshi)

vs

| Reviews | Count | Reviewers |
| --- | ----- | --------- |
| **NACKs**  | 1 | [aureleoules](https://github.com/aureleoules) |
| Concept ACK | 3 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [kouloumos](https://github.com/MarcoFalke) |
| Reviews | 12 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke)|
| _Stale_ ACKs | 1 | [Satoshi](https://github.com/satoshi)

</details>


> I assume clicking on a reviewer's name takes you to their review?

Yes!

> Is it too much to grasp the opportunity of mentioning code review terms to also drop somewhere in there a hyperlink to [bitcoin/bitcoin@master/CONTRIBUTING.md#decision-making-process](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-28T15%3A24%3A28Z#decision-making-process)?

Yes, could be nice to have it in the summary comment.

aureleoules replied:
@MarcoFalke I see that you're rewriting parts of the bot in Rust. Would you like me to rewrite this PR in Rust? I'm not familiar with it but would love the learning exercise!

MarcoFalke replied:
Sure, why not, if you want to

aureleoules commented:
I have removed the PR labeling feature and instead replaced it by an ACK tracker in a summary comment. The PR description has been updated accordingly.
The PR is now written in Rust.

MarcoFalke commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+__pycache__
+drahtbot/target
```
nit: maybe this can be renamed from drahtbot to review_summary, to indicate what it does in the name? Also, to avoid confusion with the collection of scripts with the same name.

aureleoules replied:
In the future, I think it would be nice to move the current scripts into this webhook framework which is why I named it drahtbot. I plan later to add the conflicting list of PRs in the summary comment (there's a TODO in the code).

MarcoFalke replied:
Not sure if it is useful to move everything into the webhook framework. For example conflicts will need to be re-calculated on every push to master. However, the complexity is N^2 in the number of pulls for that case, so it is simply not possible to do in a real-time fashion. Given that it is impossible in real-time, and given that the update rate needs to be limited, it might as well be done independently of a webhook.

aureleoules replied:
I see. Maybe `webhook_features` is more appropriate then? Labels for example could be moved to this webhook framework so that whenever a PR is created, labels are applied, instead of rescanning every PRs. Same for the "Needs rebase" label.

aureleoules commented:
You guys might be interested following our discussion on IRC @achow101 @jonatack @vasild @brunoerg.

MarcoFalke reviewed (commented):
left two nits. Might review later

aureleoules commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+__pycache__
```
:)
will fix the nits later

MarcoFalke commented on drahtbot/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,493 @@
+}
+
+fn summary_comment_template(initial: bool, acks: Option<Vec<Review>>) -> String {
+    let mut comment = r#"## Summary
```
Given that a synchronous update is impossible, it might be better to re-use the existing "async" comment structure. See https://github.com/MarcoFalke/DrahtBot/blob/97274a4dcd00a207d4e6906e4b8272b7493a3356/scripts/util/util.py#L15

(I'll port it to rust soon)

Otherwise every pull request might have two summary comments.

aureleoules replied:
> Given that a synchronous update is impossible, it might be better to re-use the existing "async" comment structure. See 

Right good idea.

> This could also allow to get rid of having to pass the name of the bot via command line?

Checking the bot username could make sure that a user doesn't include this hidden comment to mess with the bot no? 
Or the username could be deduced at runtime with an API request.

MarcoFalke replied:
Would be hard to mess with the bot, as the first comment is picked. Even if a user races the bot, it doesn't seem to be able to cause any harm?

MarcoFalke replied:
This would also make it easier to replace the bot if a bot user account gets blocked/deleted

MarcoFalke replied:
> (I'll port it to rust soon)

Done.

You should now be able to call `let cmt = util::get_metadata_sections(pull_number)` and then `util::update_metadata_comment(cmt, "reviews ...", util::IdComment::SecReviews)`

aureleoules replied:
I addressed your change (had to rebase), seems to work great!
If the summary comment was not already created on a PR, it will be created but won't appear at the top. This should be fine for old PRs.

vasild commented:
Concept ACK

I have the feeling that it may as well be ok to just list one ACK per line in the table, without grouping. Grouping is a kind of complication that is needed if there are a lot of entries in the table.

Anyway, if there is to be grouping then the groups should be aligned with the ones mentioned in `CONTRIBUTING.md`. For example, when I read the OP I wondered why these are two distinct groups: `ACK` and `Code Review ACK`, they are the same thing, no? Then I read that `Code Review ACK` does not require a commit id. I don't remember seen such ACKs anywhere. `CONTRIBUTING.md` categories are:

* `Concept ACK`
* `Concept NACK`
* `Approach ACK`
* `Approach NACK`
* `ACK BRANCH_COMMIT followed by a description of how the reviewer did the review`

Thanks, this looks cool :ice_cube: :fire:

aureleoules commented:
> I have the feeling that it may as well be ok to just list one ACK per line in the table, without grouping. Grouping is a kind of complication that is needed if there are a lot of entries in the table.

I think I would prefer grouping as it seems easier to read the table but I'll let others express their opinions on that.

> ACK and Code Review ACK, they are the same thing, no?

The way I understand it is that Code Review ACK means the code is well written, in terms of C++ or Python but the behavior may not have necessarily been carefully reviewed?

From [jon atack's blog](https://jonatack.github.io/articles/how-to-review-pull-requests-in-bitcoin-core):
> Reviewers sometimes comment "Code review ACK" to communicate that the code looks good but they haven't tested the changes or don't have an opinion on the concept yet.

> Then I read that Code Review ACK does not require a commit id

Right, this is just a boolean to switch in this PR code, I'll enable it.

aureleoules commented:
- Renamed `drahtbot` to `webhook_features` for potential future features that require webhooks
- Remove `__pycache__` from 	`.gitignore`
- `Code Review ACK` now requires a commit hash

https://github.com/MarcoFalke/DrahtBot/compare/b3f0d84361040aec0dc9b7d884a9317f9fe922cc..2fd112e9f8eade886fdd93f1e0276b71cb5d6181

aureleoules commented:
- Rebased
- Added test vectors for `parse_acks_in_comment`
- Fixed Code Review ACKs not being recognized
- Update existing summary comment with hidden ID as suggested by @MarcoFalke.

aureleoules commented:
Updated code to retrieve the metadata comment from an existing array of GitHub comments instead of fetching them twice.
https://github.com/MarcoFalke/DrahtBot/compare/6b3f8974e11f2cdc7d318b4ea1f1baefbae1c627..0f8767312ab54f6aa4eb22d49c027dbd7be13aa8

MarcoFalke commented on util/src/lib.rs:
```diff
@@ -133,26 +135,35 @@ pub async fn get_metadata_sections(
+) -> MetaComment {
     for c in comments {
-        let b = c.body.expect("remote api error");
+        let b = c.body.clone().expect("remote api error");
```
any reason to create a copy when a reference is enough?
```suggestion
        let b = c.body.as_ref().expect("remote api error");
```

MarcoFalke commented on util/src/lib.rs:
```diff
@@ -133,26 +135,35 @@ pub async fn get_metadata_sections(
         }
     }
-    Ok(MetaComment {
+    return MetaComment {
```
```suggestion
    MetaComment {
```

no need for return

MarcoFalke commented on webhook_features/README.md:
````diff
@@ -0,0 +1,8 @@
+
+```bash
+$ cargo build --release
+$ ./target/release/drahtbot --repo bitcoin/bitcoin --repo bitcoin-core/gui --token $GITHUB_TOKEN --bot-username DrahtBot
````
```suggestion
$ ./target/release/drahtbot --repo bitcoin/bitcoin --repo bitcoin-core/gui --token $GITHUB_TOKEN
```

can be removed now?

aureleoules replied:
I just fixed a bug where the bot would run twice because it would receive a webhook event of its own comment edition after someone comments an ACK, so it would re-run the parsing etc.
https://github.com/MarcoFalke/DrahtBot/compare/e3ba4dac07cfe6b9f65f5d645c7b13a87a9ea597..73c373b81f7d69d6982e49f60a17007b4c1e5905

So I think the bot username is needed, but can be deduced at runtime with an API req if you prefer.

aureleoules replied:
Removed `--repo` and now deduce username at runtime.

MarcoFalke reviewed (commented):
left some more nits

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+lazy_static! {
+    static ref ACK_PATTERNS: Vec<(&'static str, AckType)> = multi_vec![
+        ["code review ack", "cr ack", "cr-ack", "crack"] => AckType::CodeReviewACK;
+        ["code review nack", "cr nack", "cr-nack", "crnack"] => AckType::CodeReviewNACK;
```
Is there any example of this ever being used?

aureleoules replied:
Nope, removed

MarcoFalke reviewed (commented):
some more comments

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+
+fn summary_comment_template(initial: bool, acks: Option<Vec<Review>>) -> String {
+    let mut comment = r#"
+### ACKs
```
```suggestion
### Reviews
```

aureleoules replied:
Done

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+fn summary_comment_template(initial: bool, acks: Option<Vec<Review>>) -> String {
+    let mut comment = r#"
+### ACKs
+Please ACK this PR if you have reviewed it and believe it to be ready for merging.
```
what about linking to the review guideline, as suggested by another reviewer previously?

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+        comment += "\n";
+    }
+
+    // TODO: PR Conflicts
```
can remove?

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+            ))?
+    };
+
+    let cmt = util::get_metadata_sections(&octocrab, &octocrab.issues(owner.clone(), repo_name.clone()), pr_number).await?;
```
any reason to clone this here, when a reference can be passed instead?

aureleoules replied:
got this
```
   --> src/features/summary_comment.rs:217:71
    |
217 |     let cmt = util::get_metadata_sections(&octocrab, &octocrab.issues(&owner, &repo_name), pr_number).await?;
    |                                                                ------ ^^^^^^
    |                                                                |      |
    |                                                                |      the trait `std::convert::From<&&str>` is not implemented for `std::string::String`
    |                                                                |      help: consider dereferencing here: `&*owner`
    |                                                                required by a bound introduced by this call
    |
```

MarcoFalke replied:
The type of owner is already `&str`, see https://docs.rs/serde_json/latest/serde_json/value/enum.Value.html#method.as_str

So adding `&` will make it `&&str`, which wouldn't work

aureleoules replied:
Right, not sure why I did that, thanks.

aureleoules commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,631 @@
+        ["concept ack", "concept-ack", "conceptack"] => AckType::ConceptACK;
+        ["concept nack", "concept-nack", "conceptnack"] => AckType::ConceptNACK;
+        ["approach ack", "approach-ack", "approachack"] => AckType::ApproachACK;
+        ["approach nack", "approach-nack", "approachnack"] => AckType::ApproachNACK;
```
Since Concept NACKs now map to NACKs I think it's ok to differentiate NACKs and Approach NACKs.
I don't mind either way.

aureleoules commented:
I addressed all your comments @MarcoFalke, thanks.
I kinda messed up my rebase so sorry for the long diff but most changes are [here](https://github.com/MarcoFalke/DrahtBot/compare/73c373b81f7d69d6982e49f60a17007b4c1e5905..0aca8ee53167b0bae5e3408e3511441ce519024f).
I just added a new -debug flag to enable running the bot on the same account as the reviewer (otherwise events would get skipped).

MarcoFalke commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,633 @@
+
+lazy_static! {
+    static ref ACK_PATTERNS: Vec<(&'static str, AckType)> = multi_vec![
+        ["code review ack", "cr ack", "cr-ack", "crack"] => AckType::CodeReviewACK;
```
Why not map this to ::ACK? Sure it is possible that someone did *not* review the code (maybe only did testing in the GUI or so), but then this should be mentioned in the review comment, see https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#code-review

vasild commented on webhook_features/src/features/summary_comment.rs:
```diff
@@ -0,0 +1,633 @@
+}
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+enum AckType {
+    ACK,
+    NACK,
+    CodeReviewACK,
+    ConceptACK,
+    ApproachACK,
+    ApproachNACK,
```
It would be really nice if this list is in sync with the guidelines from https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review. If the above stays as it is, do you plan to update `CONTRIBUTING.md` to bring them in sync? Or if the above is to match `CONTRIBUTING.md` then it should be:

```diff
 enum AckType {
     ACK,
-    NACK,
-    CodeReviewACK,
     ConceptACK,
+    ConceptNACK,
     ApproachACK,
     ApproachNACK,
```

aureleoules commented:
Thanks @vasild @MarcoFalke. 
- Mapped `NACK` with `Concept NACK`
- Mapped `Code Review ACK` with `ACK`

See diff https://github.com/MarcoFalke/DrahtBot/compare/31d791e87bac08d3b3f6080138d25f938af23cec..2a9740a3d60f0b4ad2032e3253ef1987b5d786fd

MarcoFalke commented:
I've rewritten/simplified large chunks of the code after merge. It compiles, but I did not test it, so let me know if I introduced any obvious behaviour bugs

aureleoules commented:
I haven't debugged the issue but if I ACK a commit, it is detected as stale:
![image](https://user-images.githubusercontent.com/22493292/201987131-4ae1bdca-8417-4b04-8d9a-4009a76ea3ad.png)

aureleoules commented:
And ACKing an invalid commit results in an 'ACK':
![image](https://user-images.githubusercontent.com/22493292/201987334-a8fed428-4166-4c4b-b940-82532a7bc772.png)

MarcoFalke commented:
Good catch. I accidentally typed `==` when I meant to type `!=`. Fixed in 8e505354d, along with a fix to only check the prefix.

aureleoules commented:
> Fixed in https://github.com/MarcoFalke/DrahtBot/commit/8e505354d10060c1c32ece58306286cfdafd21e2, along with a fix to only check the prefix.

Nice, the bot seems to work fine now!

MarcoFalke commented:
The bot is disabled for closed stuff, so you can test here https://github.com/MarcoFalke/DrahtBot/pull/3#issuecomment-1315833299

aureleoules commented:
I noticed a bug in https://github.com/bitcoin/bitcoin/pull/26257:
The conflict script triggered an update of the summary comment, which triggered the parsing of reviews but didn't take into account the recent ACK of stickies-v, instead there is still a Stale ACK.
I assume it is because of [6d4d721e10ff8f6ec71f5b401afde88edd866830](https://github.com/MarcoFalke/DrahtBot/commit/6d4d721e10ff8f6ec71f5b401afde88edd866830#diff-2c3d5c887cbd3b68178ea876628a22d8d01e81297714c55bb469b61b2691c6ffR227) ?

MarcoFalke commented:
It is not detected as review, because it is an URL, not a commit:

```
re-ACK https://github.com/bitcoin/bitcoin/commit/fc3ca6f87f16448c4f4b9ecc3afa813f12868c18

...

aureleoules commented:
Ah you're right. Should we parse them as well?

I also found a bug here: https://github.com/bitcoin/bitcoin/pull/26503
In the summary comment, hebasto's review link points to an older ACK.

MarcoFalke commented:
>  In the summary comment, hebasto's review link points to an older ACK.

This is not a bug. It is always possible that a webhook is missed. It will be updated on the next trigger

aureleoules commented:
Alright, great.

aureleoules commented:
https://github.com/bitcoin/bitcoin/pull/26457#issuecomment-1317212604 :grimacing:

aureleoules commented:
I'll work on it. You should delete the duplicate summary comments (also here https://github.com/bitcoin/bitcoin/pull/26328).
I found a way to reproduce:
- PR should not have a summary comment
- Leave review comment on a line, click start review and submit
- It fires 2 events

```
Handling: aureleoules/test pull_request_review::edited
Handling: aureleoules/test pull_request_review::submitted
```

aureleoules commented:
Fixed in #4, can be verified with the steps above this comment.
